README.md
# High Level App Description

1. `main.js`: This file contains the main code for the licensing server. It creates an Express.js app, sets up routes for API endpoints, and listens on a specified port. It also handles command line arguments.
- `express`: This is the main function that creates an instance of the Express.js app.
- `slowDown`: This is a middleware function that slows down requests to the server to prevent abuse.
- `pool`: This is a connection pool for the MariaDB database.
- `handleCommandLine`: This function handles command line arguments passed to the server.
- `activateLicense`: This function activates a license for a given product.
- `licenseManager`: This is an object that contains functions for managing licenses.
- `app.get`: This sets up a GET route for the `/ping` endpoint.
- `app.post`: This sets up a POST route for the `/api/activate/:license` endpoint.
- `app.listen`: This starts the server and listens on a specified port.

2. `licenseManager.js`: This file contains functions for managing licenses. It has functions for generating, retrieving, updating, and changing the state of licenses. It also has a function for checking the state of a license.
- `generateLicense`: This function generates a new license for a given product.
- `getLicense`: This function retrieves a license for a given product.
- `updateLicense`: This function updates a license for a given product.
- `changeLicenseState`: This function changes the state of a license for a given product.
- `checkLicenseState`: This function checks the state of a license for a given product.
- `getAllLicenses`: This function retrieves all licenses for all products.

3. `productManager.js`: This file contains functions for managing products. It has functions for creating and retrieving products.
- `createProduct`: This function creates a new product.
- `getProduct`: This function retrieves a product by its ID.

INFORMATION/CHANGELOG.md
v0.0.5
Updated createTable.js to also create licences_db and user.
Implemented dotenv module.
v0.0.4
Added error handling on all functions and added end point to verify license was active.
Added counter measure to brute force attacks.
Added more columns in the license db to support future use cases.
v0.0.3
Added end point to activate license
v0.0.2
Modulirized all app for easy mainantenance
v0.0.1
Initial release

testdb.js
const mariadb = require('mariadb');

// Replace the values with your own credentials
const DB_HOST = 'localhost';
const DB_USER = 'admin_license_server';
const DB_PASSWORD = 'K&vKEi87a2g';
const DB_NAME = 'licenses_db';

const pool = mariadb.createPool({
  host: DB_HOST,
  user: DB_USER,
  port: 3080,
  password: DB_PASSWORD,
  database: DB_NAME
});

async function createTableAndInsertData() {
  let conn;
  try {
    conn = await pool.getConnection();
    
    // Create the `users` table
    await conn.query(`
      CREATE TABLE IF NOT EXISTS users (
        id INT PRIMARY KEY AUTO_INCREMENT,
        name VARCHAR(255) NOT NULL,
        email VARCHAR(255) NOT NULL UNIQUE,
        age INT
      )
    `);
    console.log('Table created.');

    // Insert some dummy data
    await conn.query(`
      INSERT INTO users (name, email, age)
      VALUES
        ('John', 'john@example.com', 30),
        ('Jane', 'jane@example.com', 25),
        ('Bob', 'bob@example.com', 40)
    `);
    console.log('Data inserted.');
    
  } catch (err) {
    console.error(err);
  } finally {
    if (conn) return conn.end();
  }
}

createTableAndInsertData();


createTable.js
const pool=require("./db");
pool.getConnection().then(e=>{
  e.query(`
    CREATE TABLE IF NOT EXISTS products (
      id VARCHAR(36) PRIMARY KEY,
      name VARCHAR(50) NOT NULL UNIQUE,
      creation_date DATETIME NOT NULL
    )
  `).then(()=>{
    console.log("products table created")
  }).catch(e=>{
    console.error("Error creating products table:",e)
  });

  e.query(`
    CREATE TABLE IF NOT EXISTS licenses (
      id INT PRIMARY KEY AUTO_INCREMENT,
      license_key VARCHAR(36) NOT NULL UNIQUE,
      product_id VARCHAR(36) NOT NULL,
      state ENUM("inactive", "active", "expired") NOT NULL,
      expiration_date DATETIME NOT NULL,
      activation_date DATETIME,
      creation_date DATETIME NOT NULL,
      license_type ENUM('trial', 'standard', 'enterprise') NOT NULL,
      user_email VARCHAR(255),
      purchase_date DATETIME,
      renewal_date DATETIME,
      max_activations INT,
      current_activations INT,
      notes TEXT,
      machine_id VARCHAR(50),
      friendly_name VARCHAR(100),
      FOREIGN KEY (product_id) REFERENCES products(id) ON DELETE CASCADE
    )
  `).then(()=>{
    console.log("licenses table created")
  }).catch(e=>{
    console.error("Error creating licenses table:",e)
  });

  e.release()
}).catch(e=>{
  console.error("Error creating tables:",e)
});

commandHandler.js
const {
    createProduct,
    showProducts
} = require("./productManager"), {
    generateLicense,
    renovateLicense,
    changeLicenseState
} = require("./licenseManager"), showLicenses = require("./licenseManager")["showLicenses"];

function handleCommandLine(e) {
    switch (e[0]) {
        case "createProduct":
            e[1] ? createProduct(e[1]) : console.log("Usage: createProduct <productName>");
            break;
        case "showProducts":
            showProducts();
            break;
        case "generateLicense":
            e[1] && e[2] ? generateLicense(e[1], parseInt(e[2])) : console.log("Usage: generateLicense <productName> <days>");
            break;
        case "renovateLicense":
            e[1] && e[2] ? renovateLicense(e[1], parseInt(e[2])) : console.log("Usage: renovateLicense <licenseKey> <days>");
            break;
        case "changeLicenseState":
            e[1] && e[2] ? changeLicenseState(e[1], e[2]) : console.log("Usage: changeLicenseState <licenseKey> <newState>");
            break;
        case "show":
            e[1] ? showLicenses(e[1]) : console.log("Usage: show <product>");
            break;
        case "help":
            handleHelp();
            break;
        default:
            console.log('Invalid command. Type "help" to view available commands and their usages.')
    }
}

function handleHelp() {
    console.log(`
Available commands:
- createProduct <productName>
- showProducts
- generateLicense <productName> <days>
- renovateLicense <licenseKey> <days>
- changeLicenseState <licenseKey> <newState>
- show <product>
- help
- exit
`)
}
module.exports = {
    handleCommandLine: handleCommandLine
};

licenseManager.js
const uuidv4 = require("uuid").v4,
  pool = require("./db");


  async function checkLicenseState(licenseKey) {
    const connection = await pool.getConnection();
    const [license] = await connection.query("SELECT * FROM licenses WHERE license_key = ?", [licenseKey]);
    connection.release();
  
    if (license) {
      return { status: 200, message: license.state };
    } else {
      return { status: 404, message: "License not found." };
    }
  }
  

  

  async function generateLicense(productName, days) {
    try {
      const connection = await pool.getConnection();
      const [product] = await connection.query("SELECT * FROM products WHERE name = ?", [productName]);
  
      if (product) {
        const licenseKey = uuidv4();
        const productId = product.id;
        const creationDate = new Date();
        let expirationDate = new Date(creationDate);
        expirationDate.setDate(creationDate.getDate() + days);
  
        await connection.query(
          "INSERT INTO licenses (license_key, product_id, state, expiration_date, creation_date) VALUES (?, ?, ?, ?, ?)",
          [licenseKey, productId, "inactive", expirationDate, creationDate]
        );
  
        console.log(`License generated for product "${productName}" with key ${licenseKey} and expiration date ` + expirationDate.toISOString());
        connection.release();
      } else {
        console.error("Product not found: " + productName);
      }
    } catch (err) {
      console.error("Error generating license:", err);
    }
  }


async function getLicense(e) {
  try {
    var n = await pool.getConnection(),
      [t] = await n.query("SELECT * FROM licenses WHERE license_key = ?", [e]);
    return t ? (n.release(), t) : (console.error("License not found: " + e), null)
  } catch (e) {
    return console.error("Error retrieving license:", e), null
  }
}
async function renovateLicense(e, n) {
  try {
    var t, r = await pool.getConnection(),
      [a] = await r.query("SELECT * FROM licenses WHERE license_key = ?", [e]);
    a ? (new Date, (t = new Date(a.expiration_date)).setDate(t.getDate() + n), await r.query("UPDATE licenses SET expiration_date = ? WHERE license_key = ?", [t, e]), console.log(`License ${e} updated. New expiration date: ` + t.toISOString()), r.release()) : console.error("License not found: " + e)
  } catch (e) {
    console.error("Error renovating license:", e)
  }
}
async function changeLicenseState(e, n) {
  try {
    var t = await pool.getConnection(),
      [r] = await t.query("SELECT * FROM licenses WHERE license_key = ?", [e]);
    r ? (await t.query("UPDATE licenses SET state = ? WHERE license_key = ?", [n, e]), console.log(`License ${e} state updated. New state: ` + n), t.release()) : console.error("License not found: " + e)
  } catch (e) {
    console.error("Error changing license state:", e)
  }
}


async function activateLicense(licenseKey, machineId) {
  const connection = await pool.getConnection();
  const [license] = await connection.query("SELECT * FROM licenses WHERE license_key = ?", [licenseKey]);

  if (license) {
    if (license.state !== "inactive") {
      connection.release();
      return { status: 400, message: `License ${licenseKey} is not inactive.` };
    } else if (new Date(license.expiration_date) < new Date()) {
      connection.release();
      return { status: 400, message: `License ${licenseKey} has already expired.` };
    } else {
      const activationDate = new Date();
      await connection.query("UPDATE licenses SET state = 'active', machine_id = ?, activation_date = ? WHERE license_key = ?", [machineId, activationDate, licenseKey]);
      connection.release();
      return { status: 200, message: `License ${licenseKey} activated for product ${license.product_id}.` };
    }
  } else {
    connection.release();
    return { status: 404, message: `License ${licenseKey} not found.` };
  }
}

async function deactivateLicense(licenseKey) {
  const connection = await pool.getConnection();
  const [license] = await connection.query("SELECT * FROM licenses WHERE license_key = ?", [licenseKey]);

  if (!license) {
    connection.release();
    return { status: 404, message: `License ${licenseKey} not found.` };
  }

  if (license.state !== 'active') {
    connection.release();
    return { status: 400, message: `License ${licenseKey} is not active.` };
  }

  await connection.query("UPDATE licenses SET state = 'inactive', machine_id = NULL WHERE license_key = ?", [licenseKey]);
  connection.release();

  return { status: 200, message: `License ${licenseKey} deactivated.` };
}


async function getAllLicenses() {
  try {
    var e = await pool.getConnection(),
      [n] = await e.query("SELECT * FROM licenses");
    return e.release(), n
  } catch (e) {
    return console.error("Error retrieving all licenses:", e), []
  }
}
async function showLicenses(e) {
  try {
    var n, t = await pool.getConnection(),
      [r] = await t.query("SELECT * FROM products WHERE name = ?", [e]);
    r ? (n = await t.query("SELECT * FROM licenses WHERE product_id = ?", [r.id]), console.log(`Licenses for product "${e}":`), console.table(n)) : console.error("Product not found: " + e), t.release()
  } catch (e) {
    console.error("Error getting licenses for product:", e)
  }
}

const machineIdRegex = /^[0-9a-zA-Z]{4}-[0-9a-zA-Z]{4}-[0-9a-zA-Z]{4}-[0-9a-zA-Z]{4}-[0-9a-zA-Z]{4}$/;
async function validateMachineId(machineId) {
  return machineIdRegex.test(machineId);
}


module.exports = {
  generateLicense: generateLicense,
  getLicense: getLicense,
  renovateLicense: renovateLicense,
  changeLicenseState: changeLicenseState,
  activateLicense: activateLicense,
  deactivateLicense : deactivateLicense,
  getAllLicenses: getAllLicenses,
  showLicenses: showLicenses
};

productManager.js
const uuidv4 = require("uuid").v4;
const pool = require("./db");

async function createProduct(productName) {
  try {
    const connection = await pool.getConnection();
    const productId = uuidv4();
    const creationDate = new Date();

    await connection.query("INSERT INTO products (id, name, creation_date) VALUES (?, ?, ?)", [productId, productName, creationDate]);
    console.log(`Product "${productName}" created with ID ${productId}.`);
    connection.release();
  } catch (err) {
    if (err.code === "ER_DUP_ENTRY") {
      console.error(`Error creating product: Product "${productName}" already exists.`);
    } else {
      console.error("Error creating product:", err);
    }
  }
}

async function showProducts() {
  try {
    var o = await pool.getConnection(),
      r = await o.query("SELECT * FROM products");
    console.log("All products:"), console.table(r), o.release()
  } catch (o) {
    console.error("Error getting products:", o)
  }
}
module.exports = {
  createProduct: createProduct,
  showProducts: showProducts
};

main.js
// Import required modules
const express = require("express");
const slowDown = require("express-slow-down");
const pool = require("./db");
const handleCommandLine = require("./commandHandler").handleCommandLine;
const activateLicense = require("./licenseManager").activateLicense;
const getLicense = require("./licenseManager").getLicense;
const licenseManager = require("./licenseManager");

// Configure the app and port
const app = express();
const port = 6000;

// Configure the slow down middleware
const speedLimiter = slowDown({
  windowMs: 20 * 1000, // 20 seconds
  delayAfter: 5, // Allow 5 requests without delaying
  delayMs: 5000, // Add 1000ms delay per request above delayAfter
});

// Apply the slow down middleware to all routes
app.use(speedLimiter);

// Enable JSON parsing for the request body
app.use(express.json());

// API endpoint to activate a license
const activateLicenseAPI = async (req, res) => {
  const licenseKey = req.params.license;
  const machineId = req.params.machine_id;
  const result = await activateLicense(licenseKey, machineId);
  res.status(result.status).send(result.message);
};

// API endpoint to deactivate a license
const deactivateLicenseAPI = async (req, res) => {
  const licenseKey = req.params.license;
  const machineId = req.body.machineId;
  const result = await licenseManager.deactivateLicense(licenseKey, machineId);
  res.status(result.status).send(result.message);
};

// API endpoint to get all licenses
app.get("/api/licenses", async (req, res) => {
  const licenses = await licenseManager.getAllLicenses();
  res.status(200).send(licenses);
});

// API endpoint to get the license status
async function getLicenseStatusAPI(req, res) {
  const licenseKey = req.params.license;
  const machineId = req.params.machine_id;

  if (!await licenseManager.validateMachineId(machineId)) {
    res.status(400).send({ status:400, message: `Invalid machine_id format: ${machineId}` });
    return;
  }

  const license = await getLicense(licenseKey);

  if (license) {
    if (license.machine_id !== machineId) {
      res.status(400).send({ status: 400, message: `License ${licenseKey} is not activated on the given machine.` });
    } else {
      res.status(200).send({ status: 200, state: license.state, expiration_date: license.expiration_date });
    }
  } else {
    res.status(404).send({ status: 404, message: "License not found." });
  }
}

// Register API routes
app.post("/api/deactivate/:license", deactivateLicenseAPI);
app.post("/api/activate/:license", activateLicenseAPI);
app.get("/api/licenses/:license", getLicenseStatusAPI);

// Endpoint for checking server connection
app.get("/ping", (req, res) => {
  res.send("pong");
});

//Deactivate ENDPOINT
app.post("/api/activate/:license/:machine_id", activateLicenseAPI);

// Start the server
app.listen(port, () => {
  console.log("Licensing Server listening at http://localhost:" + port);
});

// Connect to the database
pool
  .getConnection()
  .then(() => {
    console.log("Connected to MariaDB database.");
  })
  .catch((err) => {
    console.error("Error connecting to database:", err);
    process.exit(1);
  });

// Handle process events
process.on("unhandledRejection", (err) => {
  console.error("Unhandled rejection:", err);
  process.exit(1);
});

process.on("uncaughtException", (err) => {
  console.error("Uncaught exception:", err);
  process.exit(1);
});

process.on("SIGINT", () => {
  console.log("Exiting Licensing Server.");
  process.exit(0);
});

// Handle command line input
handleCommandLine("");
